compiler/hash-parser/src/gen.rs

```diff
@@ -177,24 +183,75 @@ where
         self.node(Name { ..*name.body() })
     }
 
-    pub(crate) fn unexpected_token_error(
+    pub(crate) fn unexpected_token_error<T>(
         &self,
         kind: &TokenKind,
         expected: &TokenKindVector,
         location: &Location,
-    ) -> ParseError {
-        ParseError::Parsing {
-            message: format!("Unexpected token '{}', expecting {}", kind, expected),
-            src: self.source_location(location),
+    ) -> ParseResult<T> {
+        // we need to convert a token-tree into just the delimiter since we don't want
+        // to print the whole tree...
+        let atom = kind.to_atom();
+
+        if expected.is_empty() {
+            self.error_with_location(format!("Unexpected token '{}'", atom), location)
+        } else {
+            self.error_with_location(
+                format!("Unexpected token '{}', expecting {}", atom, expected),
+                location,
+            )
         }
     }
 
+    /// Create an error at the current location.
+    pub fn error<T, S: Into<String>>(&self, message: S) -> ParseResult<T> {
+        Err(ParseError::Parsing {
+            message: message.into(),
+            src: self.source_location(&self.current_location()),
+        })
+    }
+
+    /// Create an error at the current location.
+    pub fn error_with_location<T, S: Into<String>>(
+        &self,
+        message: S,
+        location: &Location,
+    ) -> ParseResult<T> {
+        Err(ParseError::Parsing {
+            message: message.into(),
+            src: self.source_location(location),
+        })
+    }
+
+    pub(crate) fn expected_eof<T>(&self) -> ParseResult<T> {
+        // move onto the next token
+        self.offset.set(self.offset.get() + 1);
+
+        self.error(format!(
+            "Expected the end of a definition, but got '{}'.",
+            self.current_token().kind
+        ))
+    }
+
+    /// Create a generalised "Reached end of file..." error.
+    pub(crate) fn unexpected_eof<T>(&self) -> ParseResult<T> {
+        self.error("Unexpectedly reached the end of file.")
+    }
+
+    /// Create a generalised "Reached end of file..." error.
+    pub(crate) fn unexpected_eof_with_ctx<T>(&self, ctx: impl Into<String>) -> ParseResult<T> {
+        self.error(format!(
+            "{}: but unexpectedly reached the end of file.",
+            ctx.into()
+        ))
+    }
```

Comment:
I wonder if these messages should be in a centralised error place (maybe in reporting?). Maybe the parsing should produce errors in the form of ADTs (enum variants with data).

Reply:
Yes i think this can be done a little later once the parser gets expanded error handling. This should probably happen in a different PR.